### PR TITLE
Added support for symbolic is_contiguous

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84918
* __->__ #84829

### CODE REVIEW

File: c10/core/TensorImpl.cpp [hunk 1]
```diff
@@ -885,6 +885,33 @@ void TensorImpl::ShareExternalPointer(
...
+      if (extra_meta.strides_.at(d) == z) {
+        z *= size_d;
+      } else {
+        is_contiguous = false;
+        break;
+      }
+    }
+  }
+  return is_contiguous;
+}
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp:907]: note to reviewer: need to make sure this is synchronized with the other compute contiguous

File: c10/core/TensorImpl.h [hunk 2]
```diff
@@ -230,6 +230,9 @@ struct C10_API ExtraMeta {
   SymDimVector strides_ = {1};
   SymInt numel_ = 1;
   SymInt storage_offset_ = 0;
+  bool is_contiguous_ = true;
+  bool is_channels_last_contiguous_ = false;
+  bool is_channels_last_3d_contiguous_ = false;
```
Comment (ezyang) [inline c10/core/TensorImpl.h:235]: There isn't actually any reason to store these on EmptyMeta, since they're bool (and not SymBool).

Comment (Chillee) [reply c10/core/TensorImpl.h:235]: I think in the future we might want to make them SymBool - and I think this is a bit cleaner for now?

Comment (ezyang) [reply c10/core/TensorImpl.h:235]: okey dokey

File: c10/core/SymIntArrayRef.cpp [hunk 3]
```diff
@@ -28,7 +28,11 @@ at::IntArrayRef asIntArrayRefUnchecked(c10::SymIntArrayRef ar) {
 
 // TODO: this print is bad
 std::ostream& operator<<(std::ostream& os, SymInt s) {
-  os << "SymInt(" << s.as_int_unchecked() << ")";
+  if (s.is_symbolic()) {
+    os << "SymInt(" << s.toSymIntNodeImpl()->str() << ")";
```
Comment (wconstab) [inline c10/core/SymIntArrayRef.cpp:32]: is it safe enough to assume the node is non-null?

Comment (ezyang) [reply c10/core/SymIntArrayRef.cpp:32]: yeah, that should be an invariant

Comment (ezyang) [reply c10/core/SymIntArrayRef.cpp:32]: SymInt can default initialize to non-symbolic zero

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**